### ISSUE TITLE
'powershell' is not recognized as an internal or external command, operable program or batch file

### ISSUE THREAD

**Issue by hced** [oliverschwendener/ueli#561]
I had an issue where my application shortcuts didn't show up in ueli, despite the folder (`C:\Users\ced\AppData\Roaming\Microsoft\Windows\Start Menu\Programs`) being added in the Application Search. So I opened Settings and pressed the refresh button in Application Search, when I got an error message as seen in the screenshot.

I'm on Win 10 version 2004, using latest ueli portable version 8.10.0. I have Powershell portable version 7.1.0.0 installed at C:\Portable\PowerShell, not that this should affect anything since it's a portable install that's living apart from Window's built-in version of Powershell (and I haven't changed any environment variables as far as I remember).

![ueli_powershell_bug](https://user-images.githubusercontent.com/93350/99006814-6277ff80-2543-11eb-8787-36579b525ddf.png)

**Comment by hosseinzamaninasab** [comment]
In settings, General, click on kebab Menu and clear execution log, it worked for me.

**Comment by hced** [comment]
> In settings, General, click on kebab Menu and clear execution log, it worked for me.

This doesn't work for me. Did you get the same error message saying 'powershell' is not a recognized command?

Here's the output in ueli's debug log when starting up the app:

2020-11-24T13:04:53.577Z error: Command failed: powershell -NonInteractive -NoProfile -Command "& { [Console]::OutputEncoding = [Text.UTF8Encoding]::UTF8; $ErrorActionPreference = 'SilentlyContinue'; $applications = New-Object Collections.Generic.List[String]; 'C:\ProgramData\Microsoft\Windows\Start Menu\Programs','C:\Users\ced\AppData\Roaming\Microsoft\Windows\Start Menu','C:\Users\ced\Desktop' | %{ Get-ChildItem -Path $_ -include *.lnk, *.appref-ms, *.url, *.exe -Recurse -File | % { $applications.Add($_.FullName) } }; $result = (@{ errors = @($error | ForEach-Object { $_.Exception.Message }); applications = $applications } | ConvertTo-Json); Write-Host $result; }"
'powershell' is not recognized as an internal or external command,
operable program or batch file.

```
Error: Command failed: powershell -NonInteractive -NoProfile -Command "& { [Console]::OutputEncoding = [Text.UTF8Encoding]::UTF8; $ErrorActionPreference = 'SilentlyContinue'; $applications = New-Object Collections.Generic.List[String]; 'C:\ProgramData\Microsoft\Windows\Start Menu\Programs','C:\Users\MYUSERNAME\AppData\Roaming\Microsoft\Windows\Start Menu','C:\Users\MYUSERNAME\Desktop' | %{ Get-ChildItem -Path $_ -include *.lnk, *.appref-ms, *.url, *.exe -Recurse -File | % { $applications.Add($_.FullName) } }; $result = (@{ errors = @($error | ForEach-Object { $_.Exception.Message }); applications = $applications } | ConvertTo-Json); Write-Host $result; }"
'powershell' is not recognized as an internal or external command,
operable program or batch file.

    at ChildProcess.exithandler (child_process.js:312:12)
    at ChildProcess.emit (events.js:310:20)
    at maybeClose (internal/child_process.js:1021:16)
    at Socket.<anonymous> (internal/child_process.js:443:11)
    at Socket.emit (events.js:310:20)
    at Pipe.<anonymous> (net.js:672:12)
2020-11-24T13:05:23.046Z debug: Check for updates
2020-11-24T13:05:25.429Z debug: Update check result: update not available

```

**Comment by hosseinzamaninasab** [comment]
Yep, I got that one, I have another way that expects and hope to work for you, remove the app and download the last update. (if you didn't do that, indeed :), otherwise sorry)

**Comment by tkohlmeier** [comment]
@hced Can you run the command `powershell` from a cmd window?

**Comment by hced** [comment]
> @hced Can you run the command `powershell` from a cmd window?

Seems I can't:
```
C:\Users\ME>powershell
'powershell' is not recognized as an internal or external command,
operable program or batch file.
```
Worth mentioning is that I have, at some point, downloaded and installed a newer PowerShell version (7.1.0-rc.1) than the one shipping with Windows.

**Comment by tkohlmeier** [comment]
That's the problem.

On my system, this directory is part of the system PATH variable:

    C:\Windows\System32\WindowsPowerShell\v1.0\

If you have that directory, too, and there is a powershell.exe in it, adding it should resolve this problem.

**Comment by hced** [comment]
> That's the problem.
> 
> On my system, this directory is part of the system PATH variable:
> 
> ```
> C:\Windows\System32\WindowsPowerShell\v1.0\
> ```
> 
> If you have that directory, too, and there is a powershell.exe in it, adding it should resolve this problem.

I appended my PATH var in my previous comment just before you posted this. Yes, I have that directory in my PATH and when opening `C:\Windows\System32\WindowsPowerShell\v1.0\` I do have `powershell.exe` in there.

**Comment by tkohlmeier** [comment]
That's odd... It's even there three times.

What is the result of this command? It should list the locations of the given executable:

    where powershell

In my case just this:

    C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe

**Comment by hced** [comment]
> Yep, I got that one, I have another way that expects and hope to work for you, remove the app and download the last update. (if you didn't do that, indeed :), otherwise sorry)

@hosseinzamaninasab That did not work. N.B. I'm using **portable** version of ueli. Strangely, when downloading the portable app and deleting the old `ueli` folder, running `ueli.exe` somehow remembers my old settings ... which is a bit surprising. Isn't the purpose of portable apps to maintain preferences end configs inside the portable directory and if you delete that, everything's gone without trace?

**Comment by tkohlmeier** [comment]
There is an open issue about this: #496

**Comment by hced** [comment]
> What is the result of this command? It should list the locations of the given executable:
> 
> ```
> where powershell
> ```
```
C:\Users\ME>where powershell
'where' is not recognized as an internal or external command,
operable program or batch file.
```

**Comment by hced** [comment]
I feel totally lost right now. Am I experiencing an edge case? Not sure, because @hosseinzamaninasab reported he also got this error. I'd be curious to see what he gets when typing `where powershell` in his Command Prompt.

**Comment by tkohlmeier** [comment]
Very interesting :D

That `where` command is here: `C:\Windows\System32\where.exe`

If even that does not work, there must be something really wrong with your system... I'm afraid I'm out of ideas for now... Maybe you could try something like `sfc /scannow`.

**Comment by tkohlmeier** [comment]
Just one more thing... Did you format the output of `echo %PATH%` you posted before? If you did not, then that whitespace (new lines in this case) might be the problem. There should be no whitespace at all between the semicolons and the following path.

**Comment by hced** [comment]
> Just one more thing... Did you format the output of `echo %PATH%` you posted before? If you did not, then that whitespace (new lines in this case) might be the problem. There should be no whitespace at all between the semicolons and the following path.

Yup, I formated it because it was just one long line and I wanted to spare your eyes a bit ;-)

I worry that I'm in a complete total mess all of a sudden here... Trying to figure out why this is, not recognizing a single command. Not even the `sfc` one you mentioned earlier. Bypassing the issue by doing `C:\Windows\System32\sfc.exe /scannow` which at least lets me run it. I've also tried running Command Prompt as Administrator but no commands recognized there either. So weird, but I digress.

**Comment by hced** [comment]
UPDATE

I've 100% fixed my whole issue. I'll be damned!!

Now, this was definitely an edge case and now we can close this issue (should not @hosseinzamaninasab still report he gets the error in the screenshot).

The culprit was that one (user) env var I added a few weeks back:

`PATHEXT` with the value of `.LNK`. I added this because I read somewhere that this is recommended if you want linked binaries (some-app.lnk) to be executable without the *.lnk extension in the command line. This worked, but apparently this means it now expected all commands to end with .lnk, neglecting all .exe ones in my system folders, etc.

Changing `PATHEXT` to be `.EXE;.LNK` or what seems to be the Windows 10 defaults according to [Computer Hope](https://www.computerhope.com/issues/ch000088.htm): `.com;.exe;.bat;.cmd;.vbs;.vbe;.js;.jse;.wsf;.wsh;.msc`.

Holy cow. What a relief. (As far as I'm concerned, this issue could be closed now.)
